### PR TITLE
feat: passive runner deletion

### PR DESCRIPTION
**Pull request by yanksyoon**
Applicable spec: N/A

Rejected:
- Alert logging: The alert logging is for debugging purposes and should be setup via grafana-dashboard with datasource as Loki.

### Overview

1. Refactor the module to be step-down ordering to follow [IS-Charms contributing guide](https://github.com/canonical/is-charms-contributing-guide/blob/main/CONTRIBUTING.md).
2. Refactor `_get_ssh_connection` function to return a single working SSH connection instance. (There is no need to loop over every possible network connections and try them for each business logic).
3. Create a general `_health_check` function that encapsulates the business logic of health checking for both startup and  running health checks.
4. ~Log and alert when there is an unexpected health check result to allow more insights into the state of Openstack GitHub runners.~

### Rationale

To gain more insights into interactions with Openstack.

### Juju Events Changes

None.

### Module Changes

- `openstack_manager.py`
  - refactored to step-down function ordering
  - added `_health_check` function
  - refactored `_get_ssh_connection` function

### Library Changes

None.

### Checklist

- [x] The [charm style guide](https://juju.is/docs/sdk/styleguide) was applied
- [x] The [contributing guide](https://github.com/canonical/is-charms-contributing-guide) was applied
- [x] The changes are compliant with [ISD054 - Managing Charm Complexity](https://discourse.charmhub.io/t/specification-isd014-managing-charm-complexity/11619)
- [x] The documentation is generated using `src-docs`
- [x] The documentation for charmhub is updated.
- [x] The PR is tagged with appropriate label (`urgent`, `trivial`, `complex`)

Testing has been one on PS6
image

### CODE REVIEW

File: src/openstack_cloud/openstack_manager.py [hunk 1]
```diff
@@ -545,178 +550,272 @@ def __init__(
...
+            server_name: The Openstack server instance name.
+            timeout: Timeout in seconds to attempt connection to each available server address.
+
+        Raises:
+            _SSHError: If there was an error getting a valid SSH connection.
+
+        Returns:
+            An SSH connection to OpenStack server instance.
+        """
+        server: Server | None = conn.get_server(name_or_id=server_name)
```
Comment (yanksyoon) [inline src/openstack_cloud/openstack_manager.py:759]: Refresh server on error (e.g. network not yet attached)

File: src/openstack_cloud/openstack_manager.py [hunk 2]
```diff
@@ -545,178 +550,271 @@ def __init__(
         self._cloud_config = cloud_config
         self._github = GithubClient(token=self._config.token)
 
-    @staticmethod
-    def _get_key_path(name: str) -> Path:
-        """Get the filepath for storing private SSH of a runner.
+    def reconcile(self, quantity: int) -> int:
```
Comment (yanksyoon) [inline src/openstack_cloud/openstack_manager.py:553]: Step down ordering of methods.

File: src/openstack_cloud/openstack_manager.py [hunk 3]
```diff
@@ -545,178 +550,271 @@ def __init__(
...
+            List of OpenStack instances.
+        """
+        return [
+            instance
+            for instance in cast(list[Server], conn.list_servers())
+            if instance.name.startswith(f"{self.instance_name}-")
+        ]
+
+    @staticmethod
+    def _health_check(conn: OpenstackConnection, server_name: str, startup: bool = False) -> bool:
```
Comment (yanksyoon) [inline src/openstack_cloud/openstack_manager.py:665]: Main change

File: src/openstack_cloud/openstack_manager.py [hunk 4]
```diff
@@ -545,178 +550,271 @@ def __init__(
...
+            return False
+        if server.status not in (_INSTANCE_STATUS_ACTIVE, _INSTANCE_STATUS_BUILDING):
+            return False
+        return OpenstackRunnerManager._ssh_health_check(
+            conn=conn, server_name=server_name, startup=startup
+        )
 
     @staticmethod
-    def _ssh_health_check(server: Server) -> bool:
+    def _ssh_health_check(conn: OpenstackConnection, server_name: str, startup: bool) -> bool:
```
Comment (yanksyoon) [inline src/openstack_cloud/openstack_manager.py:699]: Main change

File: src/openstack_cloud/openstack_manager.py [hunk 5]
```diff
@@ -545,178 +550,271 @@ def __init__(
...
+
+        if RUNNER_WORKER_PROCESS in result.stdout or RUNNER_LISTENER_PROCESS in result.stdout:
+            return True
+
+        logger.error("[ALERT] Health check failed for server: %s", server_name)
+        return True
+
+    @staticmethod
+    @retry(tries=3, delay=5, max_delay=60, backoff=2, local_logger=logger)
+    def _get_ssh_connection(
```
Comment (yanksyoon) [inline src/openstack_cloud/openstack_manager.py:741]: Improved _get_ssh_connection function

**Review by yanksyoon — COMMENTED** [review]
Comments to assist in PR review

File: src/openstack_cloud/openstack_manager.py [hunk 6]
```diff
@@ -545,178 +550,271 @@ def __init__(
...
+        logger.debug("Output of `ps aux` on %s stderr: %s", server_name, result.stderr)
+        logger.debug("Output of `ps aux` on %s stdout: %s", server_name, result.stdout)
+        if not result.ok or RUNNER_STARTUP_PROCESS not in result.stdout:
+            logger.warning("List all process command failed on %s ", server_name)
+            return False
+        logger.info("Runner process found to be healthy on %s", server_name)
+        if startup:
+            return True
+
+        if RUNNER_WORKER_PROCESS in result.stdout or RUNNER_LISTENER_PROCESS in result.stdout:
```
Comment (cbartz) [inline src/openstack_cloud/openstack_manager.py:733]: Is it necessary to check for both, or is it not enough to check for the latter?

File: tests/unit/test_openstack_manager.py [hunk 7]
```diff
@@ -859,3 +858,290 @@ def test_repo_policy_config(
...
+    mock_server.addresses = {}
+    mock_connection = MagicMock()
+    mock_connection.get_server.return_value = mock_server
+
+    with pytest.raises(openstack_manager._SSHError) as exc:
+        openstack_manager.OpenstackRunnerManager._get_ssh_connection(
+            conn=mock_connection, server_name="test"
+        )
+
+    assert "Missing keyfile for server" in str(exc.getrepr())
```
Comment (cbartz) [inline tests/unit/test_openstack_manager.py:1070]: Shouldn't it be `No addresses found for OpenStack server` ?

File: tests/unit/test_openstack_manager.py [hunk 8]
```diff
@@ -859,3 +858,290 @@ def test_repo_policy_config(
...
+    monkeypatch.setattr(
+        openstack_manager, "SshConnection", MagicMock(return_value=mock_ssh_connection)
+    )
+
+    with pytest.raises(openstack_manager._SSHError) as exc:
+        openstack_manager.OpenstackRunnerManager._get_ssh_connection(
+            conn=mock_connection, server_name="test"
+        )
+
+    assert "Missing keyfile for server" in str(exc.getrepr())
```
Comment (cbartz) [inline tests/unit/test_openstack_manager.py:1115]: Also here the exc msg seems not right

**Review by cbartz — COMMENTED** [review]
Minor comments and a question. Thanks!

File: src/openstack_cloud/openstack_manager.py (hunk 6 above)
Comment (yanksyoon) [reply src/openstack_cloud/openstack_manager.py:733]: Only the latter could be checked - but this has been included to match the business logic that was discussed. Later on, this constant can help determine whether the runner is busy or not :)

**Review by cbartz — APPROVED** [review]

**Review by amandahla — APPROVED** [review]
LGTM

**Pull request merged**